chapter_recurrent-modern/machine-translation-and-dataset.ipynb

```diff
@@ -0,0 +1,628 @@
+    "    StringBuilder out = new StringBuilder();\n",
+    "    Character currChar;\n",
+    "    for (int i = 0; i < text.length(); i++) {\n",
+    "        currChar = text.charAt(i);\n",
+    "        if (i > 0 && noSpace(currChar, text.charAt(i - 1))) {\n",
+    "            out.append(' ');\n",
+    "        }\n",
+    "        out.append(currChar);\n",
+    "    }\n",
+    "    return out;\n",
+    "}\n",
+    "\n",
+    "public static boolean noSpace(Character currChar, Character prevChar) {\n",
+    "    /* Preprocess the English-French dataset. */\n",
+    "    return new HashSet<>(Arrays.asList(',', '.', '!', '?')).contains(currChar)\n",
+    "            && prevChar != ' ';\n",
+    "}\n",
+    "\n",
+    "StringBuilder text = preprocessNMT(rawText.toString());\n",
+    "System.out.println(text.substring(0, 80));"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 8
+   },
+   "source": [
+    "## Tokenization\n",
+    "\n",
+    "Different from character-level tokenization\n",
+    "in :numref:`sec_language_model`,\n",
+    "for machine translation\n",
+    "we prefer word-level tokenization here\n",
+    "(state-of-the-art models may use more advanced tokenization techniques).\n",
+    "The following `tokenizeNMT` function\n",
+    "tokenizes the the first `numExamples` text sequence pairs,\n",
+    "where\n",
+    "each token is either a word or a punctuation mark.\n",
+    "This function returns\n",
+    "two lists of token lists: `source` and `target`.\n",
+    "Specifically,\n",
+    "`source.get(i)` is a list of tokens from the\n",
+    "$i^\\mathrm{th}$ text sequence in the source language (English here) and `target.get(i)` is that in the target language (French here).\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static Pair<ArrayList<String[]>, ArrayList<String[]>> tokenizeNMT(\n",
+    "        String text, Integer numExamples) {\n",
+    "    ArrayList<String[]> source = new ArrayList<>();\n",
+    "    ArrayList<String[]> target = new ArrayList<>();\n",
+    "\n",
+    "    int i = 0;\n",
+    "    for (String line : text.split(\"\\n\")) {\n",
+    "        if (numExamples != null && i > numExamples) {\n",
+    "            break;\n",
+    "        }\n",
+    "        String[] parts = line.split(\"\\t\");\n",
+    "        if (parts.length == 2) {\n",
+    "            source.add(parts[0].split(\" \"));\n",
+    "            target.add(parts[1].split(\" \"));\n",
+    "        }\n",
+    "        i += 1;\n",
+    "    }\n",
+    "    return new Pair<>(source, target);\n",
+    "}\n",
+    "\n",
+    "Pair<ArrayList<String[]>, ArrayList<String[]>> pair = tokenizeNMT(text.toString(), null);\n",
+    "ArrayList<String[]> source = pair.getKey();\n",
+    "ArrayList<String[]> target = pair.getValue();\n",
+    "for (String[] subArr : source.subList(0, 6)) System.out.println(Arrays.toString(subArr));"
```

Comment:
in general, I don't recommend to write single-line statement

Reply:
Yeah makes sense, I changed it.